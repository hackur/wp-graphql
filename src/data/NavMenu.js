import { toGlobalId } from 'graphql-relay';
import Dataloader from 'dataloader';
import fetchData from 'data/utils';

// Dataloader expects IDs that can be read by the REST API

const path = process.env.WP_NAV_MENUS_ENDPOINT || null;
if (!path) {
  throw Error(
    'This endpoint does not exist in WordPress yet. ' +
      'You must install the WordPres GraphQL Middleware plugin.'
  );
}
const navMenuLoader = new Dataloader(ids =>
  fetchData(path)
    .then(({ data: { body } }) => body)
    .then(menus => ids.map(id => menus.find(item => item.id === parseInt(id, 10))))
);

class NavMenu {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return path;
  }

  static async load(id) {
    const data = await navMenuLoader.load(id);
    return data ? Object.assign(new NavMenu(), data) : null;
  }
}

export default NavMenu;
