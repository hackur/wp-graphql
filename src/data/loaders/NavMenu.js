import Dataloader from 'dataloader';
import fetchData from 'data/utils';
import NavMenu from 'data/NavMenu';

const endpoint = NavMenu.getEndpoint();

// Dataloader expects IDs that can be read by the REST API

export default function getNavMenuLoaders() {
  const navMenuLoader = new Dataloader(ids =>
    fetchData(endpoint)
      .then(({ data: { body } }) => body)
      .then(menus => ids.map(id => menus.find(item => item.id === parseInt(id, 10))))
  );

  return {
    load: async id => {
      const data = await navMenuLoader.load(id);
      return data ? Object.assign(new NavMenu(), data) : null;
    },
  };
}
