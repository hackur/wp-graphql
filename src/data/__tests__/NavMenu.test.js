import NavMenu from 'data/NavMenu';

describe('Test NavMenu data access', () => {
  test('Get endpoint', () => {
    expect(NavMenu.getEndpoint()).toMatchSnapshot();
  });

  test('Load a NavMenu', async () => {
    const menu = await NavMenu.load('main-nav');
    expect(menu.getID()).toMatchSnapshot();
    expect(menu).toMatchSnapshot();
  });
});
