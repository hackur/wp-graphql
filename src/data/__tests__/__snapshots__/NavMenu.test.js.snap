// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NavMenu data access Get endpoint 1`] = `"graphql/v1/nav-menus"`;

exports[`Test NavMenu data access Load a NavMenu 1`] = `"TmF2TWVudTo="`;

exports[`Test NavMenu data access Load a NavMenu 2`] = `
NavMenu {
  "slug": "main-nav",
}
`;
